File: Medium-Array/SetMatrixZero.java
// WAP to set matrix zero
import java.util.Scanner;

class SetMatrixZero {
    public static void setMatrixZeroes(int[][] mat) {
        int m = mat.length;
        int n = mat[0].length;

        boolean[] row = new boolean[m];
        boolean[] col = new boolean[n];

        // Mark the rows and columns that need to be set to 0
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (mat[i][j] == 0) {
                    row[i] = true;
                    col[j] = true;
                }
            }
        }

        // Set cells to 0 based on marked rows and columns
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (row[i] || col[j]) {
                    mat[i][j] = 0;
                }
            }
        }
    }

    // Utility method to print the matrix
    public static void printMatrix(int[][] mat) {
        for (int[] row : mat) {
            for (int val : row) {
                System.out.print(val + " ");
            }
            System.out.println();
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter the number of rows in the matrix: ");
        int m = sc.nextInt();

        System.out.print("Enter the number of columns in the matrix: ");
        int n = sc.nextInt();

        int[][] mat = new int[m][n];

        System.out.println("Enter the elements of the matrix: ");
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                mat[i][j] = sc.nextInt();
            }
        }

        System.out.println("Original Matrix:");
        printMatrix(mat);

        setMatrixZeroes(mat);

        System.out.println("Matrix after setting rows and columns to zero:");
        printMatrix(mat);

        sc.close();
    }
}

File: Medium-Array/SetMatrixZero.py
# WAP to Given a matrix if an element in the matrix is 0 then you will have to set its entire column and row to 0 and then return the matrix..

# function to set the entire row and column to 0 if an element is 0
def setZeroes(matrix):
    # Get the number of rows and columns in the matrix
    rows = len(matrix)
    cols = len(matrix[0])

    # Create sets to keep track of rows and columns that need to be set to 0
    zero_rows = set()
    zero_cols = set()

    # Iterate through the matrix to find the positions of 0 elements
    for i in range(rows):
        for j in range(cols):
            if matrix[i][j] == 0:
                zero_rows.add(i)
                zero_cols.add(j)

    # Set the entire row and column to 0 for each position found
    for row in zero_rows:
        for j in range(cols):
            matrix[row][j] = 0

    # Set the entire column to 0 for each position found
    for col in zero_cols:
        for i in range(rows):
            matrix[i][col] = 0
    
    # Return the modified matrix
    return matrix

# Example usage
arr=[[1, 2, 3], [4, 0, 6], [7, 8, 9]]
result = setZeroes(arr)
print("Modified matrix:")
for row in result:
    print(row) 